OneTimePassword/Keychain.swift

```diff
@@ -2,13 +2,103 @@
+//  copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+//  SOFTWARE.
 //
 
 import Foundation
 
-func addKeychainItemWithAttributes(attributes: NSDictionary) -> NSData? {
+let kOTPService = "me.mattrubin.onetimepassword.token"
+
+private extension Token {
+    var keychainAttributes: [String: AnyObject]? {
+        guard let url = Token.URLSerializer.serialize(self),
+            let data = url.absoluteString.dataUsingEncoding(NSUTF8StringEncoding) else {
+                return nil
+        }
+        return [
+            kSecAttrGeneric as String:  data,
+            kSecValueData as String:    generator.secret,
+            kSecAttrService as String:  kOTPService,
+        ]
+    }
+}
+
+private extension PersistentToken {
+    private init?(keychainDictionary: NSDictionary) {
+        guard let urlData = keychainDictionary[kSecAttrGeneric as String] as? NSData,
+            let string = NSString(data: urlData, encoding:NSUTF8StringEncoding),
+            let secret = keychainDictionary[kSecValueData as String] as? NSData,
+            let keychainItemRef = keychainDictionary[kSecValuePersistentRef as String] as? NSData,
+            let url = NSURL(string: string as String),
+            let token = Token.URLSerializer.deserialize(url, secret: secret) else {
+                return nil
+        }
+        self.init(token: token, identifier: keychainItemRef)
+    }
+}
+
+public class Keychain {
+    public static let sharedInstance = Keychain()
+
+    public func persistentTokenWithIdentifier(identifier: NSData) -> PersistentToken? {
+        guard let result = keychainItemForPersistentRef(identifier) else {
+            return nil
+        }
+        return PersistentToken(keychainDictionary: result)
+    }
+
+    public func allPersistentTokens() -> [PersistentToken]? {
+        guard let keychainItems = allKeychainItems() as? [NSDictionary] else {
+            return nil
+        }
+        return keychainItems.flatMap({ PersistentToken(keychainDictionary: $0) })
+    }
+
+    public func addToken(token: Token) -> PersistentToken? {
+        guard let attributes = token.keychainAttributes,
+            persistentRef = addKeychainItemWithAttributes(attributes) else {
+                return nil
+        }
+        return PersistentToken(token: token, identifier: persistentRef)
+    }
+
+    public func updatePersistentToken(persistentToken: PersistentToken, withToken token: Token) -> PersistentToken? {
```

Comment:
Length Violation: Line should be 100 characters or less: currently 117 characters